src/components/AddressInput.vue

```diff
@@ -0,0 +1,394 @@
+            value = AddressInput._stripWhitespace(value)
+                .replace(/^N?Q?/, 'NQ') // enforce NQ at the beginning
+                .replace(/.{4}/g, (match, offset) => `${match}${(offset + 4) % 12 ? ' ' : '\n'}`) // form blocks
+                .substring(0, AddressInput.ADDRESS_MAX_LENGTH); // discarding the new line after last block
+
+            if (value.endsWith(' ')) {
+                // The word spacing set via css is only applied to spaces that are actually between words which is not
+                // the case for an ending space and the caret after an ending space therefore gets rendered at the wrong
+                // position. To avoid that we add a zero-width space as an artificial word. We do not add that to the
+                // template returned to input-format though to avoid it being interpreted as a typed character which
+                // would place the caret after the zero width space.
+                value += '\u200B';
+            }
+        }
+        return {
+            text: value,
+            template: 'xxxx xxxx xxxx\nxxxx xxxx xxxx\nxxxx xxxx xxxx', // used by input-format to position caret
+        };
+    }
+
+    private static _stripWhitespace(value: string) {
+        return value.replace(/\s|\u200B/g, ''); // normal whitespace, tabs, newlines or zero-width whitespace
+    }
+
+    // value that can be bound to via v-model
+    @Prop({
+        type: String,
+        default: '',
+    })
+    public value!: string;
+
+    public $refs: { textarea: HTMLTextAreaElement };
+
+    private currentValue: string = '';
+    private selectionStartBlock: number = -1;
+    private selectionEndBlock: number = -1;
+
+    private mounted() {
+        // trigger initial value change. Not using immediate watcher as it already fires before mounted.
+        this._onExternalValueChange();
+
+        // Bind selectionchange event handler. It has to be registered on document and is unfortunately not fired for
+        // selections in textareas in Firefox. Therefore we also bind the listener to focus, blur, select, click.
+        this._updateSelection = this._updateSelection.bind(this);
+        document.addEventListener('selectionchange', this._updateSelection);
+    }
+
+    private destroyed() {
+        document.removeEventListener('selectionchange', this._updateSelection);
+    }
+
+    @Watch('value')
+    private _onExternalValueChange() {
+        if (AddressInput._stripWhitespace(this.value) === AddressInput._stripWhitespace(this.currentValue)) return;
+
+        // could also be using format-input's parse and format helpers that preserve caret position but as we're not
+        // interested in that, we calculate the formatted value manually
+        let parsedValue = '';
+        this.value.split('').forEach((char) =>
+            parsedValue = parsedValue + (AddressInput._parse(char, parsedValue) || ''));
```

Comment:
You could also do
```
const parsedValue = this.value.split('').map(...).join('');
```

Reply:
Hum, not `map` but `reduce` is the nicest way to go here.
I changed it accordingly.